kit/torch/transforms/mixup.py

```diff
@@ -0,0 +1,170 @@
+from __future__ import annotations
+from enum import Enum, auto
+from typing import cast, overload
+
+import torch
+from torch import Tensor
+import torch.distributions as td
+import torch.nn.functional as F
+
+from kit.misc import str_to_enum
+
+__all__ = [
+    "BernoulliMixUp",
+    "BetaMixUp",
+    "MixUpMode",
+    "RandomMixUp",
+    "UniformMixUp",
+]
+
+
+class MixUpMode(Enum):
+
+    linear = auto()
+    geometric = auto()
+
+
+class RandomMixUp:
+    r"""Apply MixUp augmentation to a batch of tensors.
+
+    Implemention for `mixup: BEYOND EMPIRICAL RISK MINIMIZATION`.
+    This implementation allows for transformation of the the input in the absence
+    of labels -- this is relevant for contrastive methods that use mixup to generate
+    different views of samples for instance-distance discrimination -- and additionally
+    allows for the interchange of different lambda samplers and of different methods for
+    mixing up samples (linear vs geometric) based on lambda.
+
+    Note:
+        This implementation randomly mixes images within a batch.
+    """
+
+    def __init__(
+        self,
+        lambda_sampler: td.Beta | td.Uniform | td.Bernoulli,
+        mode: MixUpMode | str = MixUpMode.linear,
+        p: float = 1.0,
+        num_classes: int | None = None,
+    ) -> None:
+        super().__init__()
+        self.lambda_sampler = lambda_sampler
+        if not 0 <= p <= 1:
+            raise ValueError("'p' must be in the range [0, 1].")
+        self.p = p
+        if isinstance(mode, str):
+            mode = str_to_enum(str_=mode, enum=MixUpMode)
+        self.mode = mode
+        if (num_classes is not None) and num_classes < 1:
+            raise ValueError(f"{ num_classes } must be greater than 1.")
+        self.num_classes = num_classes
+
+    def _mix(self, tensor_a, *, tensor_b, lambda_) -> Tensor:
+        lambda_c = 1 - lambda_
+        if self.mode is MixUpMode.linear:
+            return lambda_ * tensor_a + lambda_c * tensor_b
+        return tensor_a ** lambda_ * tensor_b ** lambda_c
+
+    @overload
+    def transform(self, inputs: Tensor, *, targets: Tensor) -> dict[str, Tensor]:
+        ...
+
+    @overload
+    def transform(self, inputs: Tensor, *, targets: None) -> Tensor:
+        ...
+
+    def transform(self, inputs: Tensor, *, targets: Tensor | None) -> Tensor | dict[str, Tensor]:
```

Comment:
this might be nicer to return a union of dataclasses

Reply:
I of course considered that but I wanted to keep it in line with albumentations/kornia

Reply:
that said, I'm already flouting their convention with the overloads

Reply:
Okay, I replaced the dictionary with a named tuple